[DR-1937] Update ingest flight to use Azure storage tables

samanehsan opened the pull request:
Places to update:
- [x] File ingest flight
- [x] Dataset delete flight
- [x]  File delete flight
- [x] Endpoints that look up file metadata (right now the file ingest flight successfully writes the file metadata records in Azure, but the`fileService` still looks in Firestore)  [Q from Shelby: This is just to support lookupFileByPath() and lookupFileById() endpoints, right? The only fileservice method that appears to be used by the ingest flight is FIleService.fileModelFromFSItem(), which doesn't actually look anything up in firestore, right?]
- [x] File ingest bulk Flight
- [x] File Ingest Worker Flight

Load history table fix - It needs to be able to handle failed ingest files. When a file ingest fails, a file_id is not generated. So, we could only use the file id as the row key sometimes. I'm proposing the to use a random UUID as the row key instead. 
![image](https://user-images.githubusercontent.com/13254229/130499877-43b124f6-1b7d-4b33-87f1-54c881156dca.png)

Follow-on ticket:
- https://broadworkbench.atlassian.net/browse/DR-2051

samanehsan committed:
Update file ingest flight

samanehsan committed:
Update DatasetDeleteFlight

samanehsan committed:
Update FileDeleteFlight

samanehsan committed:
Fix rebase

snf2ye committed:
First pass to make note of needed changes and add azure steps in worker flight

wip - notes for work

snf2ye committed:
Support azure bulk file load: only get google project for GCP bulk ingest

snf2ye committed:
Add support for file lookups by id and by path

snf2ye committed:
Singe file ingest new connected test working

pull get profile out of table dao

snf2ye committed:
Get bulk ingest flights working

Add test for directory paths

clean up

snf2ye commented on src/main/java/bio/terra/service/tabulardata/azure/AzureStorageTablePdao.java:
```diff
@@ -129,7 +131,7 @@ public void storeLoadHistory(
       StorageTableLoadHistoryEntity entity, String loadTag, Instant loadTime) {
     var model = entity.model;
-    return new TableEntity(entity.partitionKey, model.getFileId())
+    return new TableEntity(entity.partitionKey, UUID.randomUUID().toString())
```
TODO: determine what we want to set this to since file id is sometimes null. Currently discussing in slack.

snf2ye replied:
@tlangs  - Did you get a chance to look at this change? Description in old slack message: https://broadinstitute.slack.com/archives/C01VBGH431S/p1629760679215600

tlangs replied:
Hmm, that's annoying, since StorageTables use the row id for ordering, and we want the ordering to be consistent between GCP and Azure. The hash of the target path might be a better idea.

tlangs replied:
Ticket created here: https://broadworkbench.atlassian.net/browse/DR-2074

snf2ye committed:
load history table

snf2ye committed:
fix file delete by switching into single transactions for azure storage tables

snf2ye commented on src/main/java/bio/terra/service/filedata/azure/tables/TableDirectoryDao.java:
```diff
@@ -135,22 +133,19 @@ public boolean deleteDirectoryEntry(TableServiceClient tableServiceClient, Strin
-          new ListEntitiesOptions()
-              .setFilter(
-                  String.format(
-                      "path eq '%s'", fileMetadataUtils.makePathFromLookupPath(lookupPath)));
+          new ListEntitiesOptions().setFilter(String.format("path eq '%s'", filterPath));
       PagedIterable<TableEntity> entities = tableClient.listEntities(options, null, null);
-      int size = List.of(entities).size();
-      if (size > 1) {
+      int size = (int) entities.stream().count();
+      if (size > 0) {
         break;
       }
-      TableEntity entity =
-          lookupByFilePath(tableServiceClient, datasetId, encodePathAsAzureRowKey(lookupPath));
-      deleteList.add(new TableTransactionAction(TableTransactionActionType.DELETE, entity));
+      TableEntity entity = lookupByFilePath(tableServiceClient, datasetId, lookupPath);
+      tableClient.submitTransaction(
+          List.of(new TableTransactionAction(TableTransactionActionType.DELETE, entity)));
       lookupPath = fileMetadataUtils.getDirectoryPath(lookupPath);
     }
-    tableClient.submitTransaction(deleteList);
```
File delete flight was failing when we had a batch transaction to delete all of the directory entries in the azure storage table. Something about trying to access different directories in the same transaction.

samanehsan replied:
Oh! We can't delete all of these rows in a single transaction because they have different partition keys. I changed this in `createDirectoryEntry` but forgot to do that here.

snf2ye replied:
Ah, ok! Thanks!

samanehsan commented on src/main/java/bio/terra/service/filedata/azure/tables/TableDirectoryDao.java:
```diff
@@ -135,22 +133,19 @@ public boolean deleteDirectoryEntry(TableServiceClient tableServiceClient, Strin
-          lookupByFilePath(tableServiceClient, datasetId, encodePathAsAzureRowKey(lookupPath));
-      deleteList.add(new TableTransactionAction(TableTransactionActionType.DELETE, entity));
+      TableEntity entity = lookupByFilePath(tableServiceClient, datasetId, lookupPath);
+      tableClient.submitTransaction(
```
Since this just deletes one row, you could simplify this by removing the transaction and calling `deleteEntity` directly:
```
if (entity != null) {
  tableClient.deleteEntity(entity);
}
```

snf2ye replied:
I didn't get to this suggestion, but I think switching to deleteEntity is a great idea!

samanehsan replied:
Done!

snf2ye replied:
Thank you!

snf2ye committed:
Allow concurrent updates to table storage directory

snf2ye committed:
Fix lookupFSItem method

Also fix the lookup by path

snf2ye committed:
Fix AzureDatasetHappyPath test by adding the client service utils

wip

wip

wip

switch to client utils

clean up

snf2ye committed:
cleanup

snf2ye commented on src/main/java/bio/terra/service/filedata/azure/tables/TableServiceClientUtils.java:
```diff
@@ -0,0 +1,35 @@
+import com.azure.data.tables.models.ListEntitiesOptions;
+import com.azure.data.tables.models.TableEntity;
+
+public class TableServiceClientUtils {
```
@giventocode  - I would definitely appreciate a review here! Maybe this is something that could live in CRL.

giventocode replied:
Sorry for the delayed response. I missed this notification. Left some comments below.

samanehsan replied:
No worries! Thanks for the review!

samanehsan committed:
Undo load history table name change

snf2ye committed:
Merge branch 'se-update-ingest-flights' of https://github.com/DataBiosphere/jade-data-repo into se-update-ingest-flights

snf2ye committed:
No longer support multi-storage accounts per dataset

snf2ye committed:
remove comments

samanehsan committed:
Check for default billing profile

For now only allow Azure file ingests
to use the default dataset billing
profile.

samanehsan committed:
Fix spotless error

samanehsan committed:
Fix unit tests mocking

samanehsan committed:
Fix azure resource config test

samanehsan committed:
Fix deleting datasets

samanehsan committed:
Remove unused code in azure tabledao

snf2ye commented on src/main/java/bio/terra/service/filedata/azure/tables/TableDirectoryDao.java:
```diff
@@ -199,6 +201,7 @@ public FireStoreDirectoryEntry retrieveByPath(
     TableClient tableClient = tableServiceClient.getTableClient(TABLE_NAME);
     ListEntitiesOptions options =
         new ListEntitiesOptions().setFilter(String.format("path eq '%s'", dirPath));
+    // TODO - add check that there are entries
```
With this TODO comment, I think I was thinking of adding the check to make sure the table exists/has entities.  But, not sure if it's really needed - totally up to you.
Something like:
`TableServiceClientUtils.tableHasEntries(tableServiceClient, TABLE_NAME)`

snf2ye commented on src/test/java/bio/terra/service/filedata/azure/tables/LoadHistoryStorageTableConnectedTest.java:
```diff
@@ -0,0 +1,158 @@
+
+  @After
+  public void cleanup() throws Exception {
+    // TODO - clean up the load history table
```
I can just create a follow-on ticket to clean this and the other connected tests up. Apologies these weren't polished before I left 😞

samanehsan committed:
Remove transaction for delete entities

nmalfroy commented on src/main/java/bio/terra/service/dataset/flight/delete/DatasetDeleteFlight.java:
```diff
@@ -59,27 +63,31 @@ public DatasetDeleteFlight(FlightMap inputParameters, Object applicationContext)
     addStep(new LockDatasetStep(datasetDao, datasetId, false, true), lockDatasetRetry);
-    addStep(new DeleteDatasetValidateStep(snapshotDao, dependencyDao, datasetService, datasetId));
     if (platform.isGcp()) {
+      // TODO: Do this check for Azure datasets
```
Is there a Jira for this? I think there is

samanehsan replied:
Good catch! I just created a ticket here: https://broadworkbench.atlassian.net/browse/DR-2070

nmalfroy commented on src/main/java/bio/terra/service/filedata/FileService.java:
```diff
@@ -187,10 +221,29 @@ FSItem lookupFSItem(String datasetId, String fileId, int depth) throws Interrupt
+      AzureStorageAccountResource storageAccountResource =
+          resourceService
+              .getStorageAccount(dataset, billingProfileModel)
+              .orElseThrow(
```
should we combine this logic (throw if not found) in the resource service?

tlangs replied:
I agree with this sentiment. No use repeating ourselves.

nmalfroy commented on src/main/java/bio/terra/service/filedata/azure/tables/TableDao.java:
```diff
@@ -0,0 +1,354 @@
+      logger.info("deleting files from dataset");
+      fileDao.deleteFilesFromDataset(tableServiceClient, func);
+    }
+    logger.info("deleting direcotry entries");
```
typo

nmalfroy reviewed (approved):
Cool!  I think this is a big enough one that we may end up iterating on it a bit more but I'd like to get this in and overall it looks great!

giventocode commented on src/main/java/bio/terra/service/filedata/azure/tables/TableServiceClientUtils.java:
```diff
@@ -0,0 +1,35 @@
+  }
+
+  public static boolean tableHasEntries(TableServiceClient tableServiceClient, String tableName) {
+    ListEntitiesOptions options = new ListEntitiesOptions();
```
One optimization here would be to set the `top` value to 1. This way you restrict the maximum number of returned entities to 1.

giventocode commented on src/main/java/bio/terra/service/filedata/azure/tables/TableServiceClientUtils.java:
```diff
@@ -0,0 +1,35 @@
+
+  public static boolean tableExists(TableServiceClient tableServiceClient, String tableName) {
+    return tableServiceClient.listTables().stream()
+        .filter(table -> table.getName().matches(tableName))
```
I think this may not be the optimal way of do it. You are returning all tables and then applying the filter.  The preference would be to have the service do the filtering. Something along these lines:

```java
ListTablesOptions options = new ListTablesOptions().setFilter("TableName eq 'myTable'");
 
 PagedIterable<TableItem> retrievedTableItems = tableServiceClient.listTables(options, Duration.ofSeconds(5),
     new Context("key1", "value1"));
 
 ```

tlangs commented on src/main/java/bio/terra/service/filedata/flight/delete/DeleteFileAzureDirectoryStep.java:
```diff
@@ -0,0 +1,53 @@
+    try {
+      boolean found =
+          tableDao.deleteDirectoryEntry(fileId, billingProfileModel, storageAccountResource);
+      DeleteResponseModel.ObjectStateEnum stateEnum =
```
Extremely minor quip: I think this could be called `state` instead of `stateEnum` since semantically, the variable corresponds to the state of the object.

tlangs commented on src/main/java/bio/terra/service/filedata/flight/ingest/IngestFileAzureDirectoryStep.java:
```diff
@@ -0,0 +1,138 @@
+            new FireStoreDirectoryEntry()
+                .fileId(fileId)
+                .isFileRef(true)
+                .path(fileMetadataUtils.getDirectoryPath(loadModel.getTargetPath()))
+                .name(fileMetadataUtils.getName(loadModel.getTargetPath()))
```
`targetPath` is already in scope for this.

tlangs commented on src/test/java/bio/terra/service/filedata/azure/AzureIngestFileConnectedTest.java:
```diff
@@ -0,0 +1,193 @@
+@Category(Connected.class)
+public class AzureIngestFileConnectedTest {
+  private static final String MANAGED_RESOURCE_GROUP_NAME = "mrg-tdr-dev-preview-20210802154510";
+  private static final String STORAGE_ACCOUNT_NAME = "tdrshiqauwlpzxavohmxxhfv";
```
this seems a little... ✨magical✨

samanehsan replied:
That's a good point...I'll see if I can use the storage account from the test configuration instead

samanehsan committed:
Optimize storage table utils

samanehsan committed:
Cleanup getStorageAccount

samanehsan commented:
> Hmm, that's annoying, since StorageTables use the row id for ordering, and we want the ordering to be consistent between GCP and Azure. The hash of the target path might be a better idea.

@tlangs do we rely on that ordering somewhere?

samanehsan committed:
More code review cleanup

samanehsan committed:
Fix unit test mocking

samanehsan committed:
Remove hard-coded azure test config resources
